src/daemon/rpmostree-sysroot-upgrader.c

```diff
@@ -1105,7 +1105,17 @@ overlay_packages (RpmOstreeSysrootUpgrader *self,
   else
     base_rev = g_strdup (self->new_revision);
 
-  if (g_strv_length (self->requested_packages) > 0)
+  /* Do we have any layered packages currently, or are we trying to
+   * add some?  If we do, then we need to check out a copy of the tree
+   * so that libdnf can read it to parse the rpmdb, load
+   * /usr/lib/os-release from it, etc.
+   *
+   * If we don't have any layered packages, let's not pay the cost of
+   * checking out another copy of the tree that we'd just toss away
+   * when doing a redeployment.
+   */
+  if (g_strv_length (self->requested_packages) > 0 ||
+      g_hash_table_size (self->packages_to_add) > 0)
```

Comment:
We already check for this optimization in `rpmostree_sysroot_upgrader_deploy()` so that we only call `overlay_packages()` if there's a final package set to compute. However, even then, there's still a chance that we hit the "trivial case" deleted below if `requested_packages - packages_to_delete = empty set`.